go/engine/saltpack_decrypt.go

```diff
@@ -113,44 +114,64 @@ func (e *SaltpackDecrypt) makeMessageInfo(me *libkb.User, mki *saltpack.MessageK
 	e.res.ReceiverIsAnon = mki.ReceiverIsAnon
 }
 
+func addToKeyring(keyring *saltpackBasic.Keyring, key *libkb.NaclDHKeyPair) {
+	keyring.ImportBoxKey((*[libkb.NaclDHKeysize]byte)(&key.Public), (*[libkb.NaclDHKeysize]byte)(key.Private))
+}
+
 // Run starts the engine.
 func (e *SaltpackDecrypt) Run(m libkb.MetaContext) (err error) {
 	defer m.CTrace("SaltpackDecrypt::Run", func() error { return err })()
 
 	// We don't load this in the --paperkey case.
 	var me *libkb.User
 
-	var key libkb.GenericKey
+	var keyring *saltpackBasic.Keyring
+	keyring = saltpackBasic.NewKeyring()
+
 	if e.arg.Opts.UsePaperKey {
 		// Prompt the user for a paper key. This doesn't require you to be
 		// logged in.
 		keypair, _, err := getPaperKey(m, nil)
 		if err != nil {
 			return err
 		}
-		key = keypair.EncryptionKey()
+		encryptionNaclKeyPair := keypair.EncryptionKey().(libkb.NaclDHKeyPair)
+		addToKeyring(keyring, &encryptionNaclKeyPair)
 	} else {
-		// Load self so that we can get device keys. This does require you to
-		// be logged in.
-		me, err = libkb.LoadMe(libkb.NewLoadUserArg(e.G()))
+		// This does require you to be logged in.
+		if !m.G().ActiveDevice.HaveKeys() {
+			return libkb.LoginRequiredError{}
+		}
+
+		// Only used in the makeMessageInfo call, which is helpful for old messages (one cannot encrypt messages with visible recipients any more).
+		me, err = libkb.LoadMe(libkb.NewLoadUserArgWithMetaContext(m))
```

Comment:
I realize this was in the original code, but can you change this for a UPAK load, rather than a full user load?  The latter is slow.

Reply:
we need the creation time of the devices to display the error message in its current form, which is not part of the upak. 
If we want to get rid of the creation time from the message we can probably do it with a UPAK, but I need to change the interface a bit. Is it ok if I put a todo and ticket this for the next iteration?
We could also get rid of this altogether, as it is only useful for messages encrypted by very old clients.

Reply:
ok, then leave it as is...